File: engine/aurora/stats.py
"""Statistics Engine — etki büyüklüğü, CI, eşleştirilmiş test, çoklu karşılaştırma düzeltmesi."""
from __future__ import annotations

from math import sqrt

import numpy as np
from scipy import stats as sps

from .models import Trial


def changes_by_condition(trials: list[Trial], outcome: str) -> dict[str, np.ndarray]:
    out: dict[str, list[float]] = {}
    for t in trials:
        if t.pre is None or t.post is None:
            continue
        out.setdefault(t.condition_id, []).append(getattr(t.post, outcome) - getattr(t.pre, outcome))
    return {k: np.asarray(v, dtype=float) for k, v in out.items()}


def mean_ci(x: np.ndarray, alpha: float = 0.05) -> tuple[float, float, float]:
    n = len(x)
    m = float(np.mean(x)) if n else float("nan")
    if n < 2:
        return m, float("nan"), float("nan")
    se = float(np.std(x, ddof=1)) / sqrt(n)
    h = sps.t.ppf(1 - alpha / 2, n - 1) * se
    return m, m - h, m + h


def cohen_d(a: np.ndarray, b: np.ndarray) -> float:
    if len(a) < 2 or len(b) < 2:
        return float("nan")
    pooled = sqrt(((len(a) - 1) * np.var(a, ddof=1) + (len(b) - 1) * np.var(b, ddof=1)) / (len(a) + len(b) - 2))
    return float((np.mean(a) - np.mean(b)) / pooled) if pooled > 0 else 0.0


def bonferroni(p: float, k: int) -> float:
    return min(1.0, p * max(1, k))


def holm(pvals: dict[str, float]) -> dict[str, float]:
    items = sorted(pvals.items(), key=lambda kv: kv[1])
    m = len(items)
    adjusted, running = {}, 0.0
    for i, (k, p) in enumerate(items):
        running = max(running, (m - i) * p)
        adjusted[k] = min(1.0, running)
    return adjusted


def analyze_study(trials: list[Trial], outcome: str, control_id: str,
                  correction: str = "bonferroni") -> dict:
    """Her koşul için: n, mean Δ, SD, %95 CI, kontrol ile Welch t, Cohen's d, düzeltilmiş p."""
    groups = changes_by_condition(trials, outcome)
    if control_id not in groups:
        raise ValueError(f"control condition {control_id!r} has no completed trials")
    ctrl = groups[control_id]
    k = max(1, len(groups) - 1)
    rows, raw_p = {}, {}
    for cid, x in groups.items():
        m, lo, hi = mean_ci(x)
        row = {"condition_id": cid, "n": int(len(x)), "mean_change": round(m, 4),
               "sd": round(float(np.std(x, ddof=1)), 4) if len(x) > 1 else None,
               "ci95": [round(lo, 4), round(hi, 4)], "cohen_d_vs_control": None, "p_value": None,
               "p_adjusted": None, "is_control": cid == control_id}
        if cid != control_id and len(x) > 1 and len(ctrl) > 1:
            t, p = sps.ttest_ind(x, ctrl, equal_var=False)
            row["cohen_d_vs_control"] = round(cohen_d(x, ctrl), 4)
            row["p_value"] = round(float(p), 5)
            raw_p[cid] = float(p)
        rows[cid] = row
    if correction == "holm":
        adj = holm(raw_p)
    else:
        adj = {cid: bonferroni(p, k) for cid, p in raw_p.items()}
    for cid, p in adj.items():
        rows[cid]["p_adjusted"] = round(p, 5)
    return {"outcome": outcome, "control_id": control_id, "correction": correction,
            "comparisons": k, "conditions": list(rows.values()),
            "note": "Effect sizes and CIs are reported; p-values alone are not a success criterion. "
                    "Results in the discovery set require replication in a validation set."}


def order_effect(trials: list[Trial], outcome: str) -> dict:
    """Sıra indeksi ile Δ arasında Spearman korelasyonu (sıra etkisi kontrolü)."""
    xs, ys = [], []
    for t in trials:
        if t.pre is not None and t.post is not None:
            xs.append(t.order_index)
            ys.append(getattr(t.post, outcome) - getattr(t.pre, outcome))
    if len(xs) < 3:
        return {"rho": None, "p": None, "n": len(xs)}
    rho, p = sps.spearmanr(xs, ys)
    return {"rho": round(float(rho), 4), "p": round(float(p), 5), "n": len(xs)}


def evidence_grade(row: dict, replicated: bool) -> str:
    """Kanıt derecesi: sadece istatistik + replikasyon. 'healing' kategorisi yok."""
    if row.get("n", 0) < 5:
        return "insufficient"
    p_adj, d = row.get("p_adjusted"), row.get("cohen_d_vs_control")
    if p_adj is None or d is None:
        return "insufficient"
    if p_adj < 0.05 and abs(d) >= 0.5:
        return "validated" if replicated else "discovery"
    if p_adj < 0.10 and abs(d) >= 0.3:
        return "weak_signal"
    return "null"

File: app/src/core/stats.ts
/** Statistics Engine — Δ, %95 CI, Welch t, Cohen d, Bonferroni. p tek başına başarı ölçütü değildir. */
import type { OutcomeKey, Trial } from "../data/types";

export interface ConditionStats {
  conditionId: string;
  n: number;
  meanChange: number;
  sd: number | null;
  ci95: [number, number] | null;
  cohenD: number | null;
  p: number | null;
  pAdjusted: number | null;
  isControl: boolean;
}

const mean = (x: number[]) => x.reduce((a, b) => a + b, 0) / x.length;
const variance = (x: number[]) => {
  const m = mean(x);
  return x.reduce((a, b) => a + (b - m) ** 2, 0) / (x.length - 1);
};

/** Öğrenci t dağılımı kritik değeri (iki yönlü α=0.05) — küçük df için tablo, büyük için normal yaklaşım */
const T975: Record<number, number> = { 1: 12.706, 2: 4.303, 3: 3.182, 4: 2.776, 5: 2.571, 6: 2.447, 7: 2.365, 8: 2.306, 9: 2.262, 10: 2.228, 12: 2.179, 15: 2.131, 20: 2.086, 25: 2.06, 30: 2.042, 40: 2.021, 60: 2.0, 120: 1.98 };
export const tCrit = (df: number): number => {
  if (df <= 0) return NaN;
  const keys = Object.keys(T975).map(Number).sort((a, b) => a - b);
  for (const k of keys) if (df <= k) return T975[k];
  return 1.96;
};

/** Düzenli gama fonksiyonu tabanlı t-dağılımı CDF (Abramowitz–Stegun yakl. yerine sayısal integral) */
export const tCdf = (t: number, df: number): number => {
  // Student t CDF via incomplete beta: I_x(df/2, 1/2), x = df/(df+t²)
  const x = df / (df + t * t);
  const ib = incompleteBeta(x, df / 2, 0.5);
  return t >= 0 ? 1 - 0.5 * ib : 0.5 * ib;
};

const logGamma = (z: number): number => {
  const g = 7;
  const c = [0.99999999999980993, 676.5203681218851, -1259.1392167224028, 771.32342877765313, -176.61502916214059, 12.507343278686905, -0.13857109526572012, 9.9843695780195716e-6, 1.5056327351493116e-7];
  if (z < 0.5) return Math.log(Math.PI / Math.sin(Math.PI * z)) - logGamma(1 - z);
  z -= 1;
  let x = c[0];
  for (let i = 1; i < g + 2; i++) x += c[i] / (z + i);
  const t = z + g + 0.5;
  return 0.5 * Math.log(2 * Math.PI) + (z + 0.5) * Math.log(t) - t + Math.log(x);
};

const incompleteBeta = (x: number, a: number, b: number): number => {
  if (x <= 0) return 0;
  if (x >= 1) return 1;
  const bt = Math.exp(logGamma(a + b) - logGamma(a) - logGamma(b) + a * Math.log(x) + b * Math.log(1 - x));
  if (x < (a + 1) / (a + b + 2)) return (bt * betacf(x, a, b)) / a;
  return 1 - (bt * betacf(1 - x, b, a)) / b;
};

const betacf = (x: number, a: number, b: number): number => {
  const MAXIT = 200, EPS = 3e-14, FPMIN = 1e-300;
  let c = 1, d = 1 - ((a + b) * x) / (a + 1);
  if (Math.abs(d) < FPMIN) d = FPMIN;
  d = 1 / d;
  let h = d;
  for (let m = 1; m <= MAXIT; m++) {
    const m2 = 2 * m;
    let aa = (m * (b - m) * x) / ((a + m2 - 1) * (a + m2));
    d = 1 + aa * d; if (Math.abs(d) < FPMIN) d = FPMIN; c = 1 + aa / c; if (Math.abs(c) < FPMIN) c = FPMIN; d = 1 / d; h *= d * c;
    aa = (-(a + m) * (a + b + m) * x) / ((a + m2) * (a + m2 + 1));
    d = 1 + aa * d; if (Math.abs(d) < FPMIN) d = FPMIN; c = 1 + aa / c; if (Math.abs(c) < FPMIN) c = FPMIN; d = 1 / d;
    const del = d * c;
    h *= del;
    if (Math.abs(del - 1) < EPS) break;
  }
  return h;
};

/** Welch t-testi, iki yönlü p */
export const welch = (a: number[], b: number[]): { t: number; df: number; p: number } => {
  const va = variance(a) / a.length, vb = variance(b) / b.length;
  const t = (mean(a) - mean(b)) / Math.sqrt(va + vb);
  const df = (va + vb) ** 2 / (va ** 2 / (a.length - 1) + vb ** 2 / (b.length - 1));
  const p = 2 * (1 - tCdf(Math.abs(t), df));
  return { t, df, p: Math.min(1, Math.max(0, p)) };
};

export const cohenD = (a: number[], b: number[]): number => {
  const pooled = Math.sqrt(((a.length - 1) * variance(a) + (b.length - 1) * variance(b)) / (a.length + b.length - 2));
  return pooled > 0 ? (mean(a) - mean(b)) / pooled : 0;
};

export const bonferroni = (p: number, k: number) => Math.min(1, p * Math.max(1, k));

export const changesByCondition = (trials: Trial[], outcome: OutcomeKey): Record<string, number[]> => {
  const out: Record<string, number[]> = {};
  for (const t of trials) {
    if (!t.pre || !t.post) continue;
    (out[t.conditionId] ??= []).push(t.post[outcome] - t.pre[outcome]);
  }
  return out;
};

export const analyzeStudy = (trials: Trial[], outcome: OutcomeKey, controlId: string): { comparisons: number; conditions: ConditionStats[] } => {
  const groups = changesByCondition(trials, outcome);
  const ctrl = groups[controlId] ?? [];
  const k = Math.max(1, Object.keys(groups).length - 1);
  const conditions = Object.entries(groups).map(([cid, x]): ConditionStats => {
    const n = x.length;
    const m = n ? mean(x) : NaN;
    const sd = n > 1 ? Math.sqrt(variance(x)) : null;
    const ci: [number, number] | null = sd !== null ? [m - (tCrit(n - 1) * sd) / Math.sqrt(n), m + (tCrit(n - 1) * sd) / Math.sqrt(n)] : null;
    let d: number | null = null, p: number | null = null;
    if (cid !== controlId && n > 1 && ctrl.length > 1) {
      d = cohenD(x, ctrl);
      p = welch(x, ctrl).p;
    }
    return { conditionId: cid, n, meanChange: m, sd, ci95: ci, cohenD: d, p, pAdjusted: p === null ? null : bonferroni(p, k), isControl: cid === controlId };
  });
  return { comparisons: k, conditions };
};

export const evidenceGrade = (r: ConditionStats, replicated = false): "insufficient" | "null" | "weak_signal" | "discovery" | "validated" => {
  if (r.n < 5 || r.pAdjusted === null || r.cohenD === null) return "insufficient";
  if (r.pAdjusted < 0.05 && Math.abs(r.cohenD) >= 0.5) return replicated ? "validated" : "discovery";
  if (r.pAdjusted < 0.1 && Math.abs(r.cohenD) >= 0.3) return "weak_signal";
  return "null";
};

/** Spearman ρ — sıra etkisi ve günlük korelasyonu için */
export const spearman = (x: number[], y: number[]): number | null => {
  if (x.length < 3 || x.length !== y.length) return null;
  const rank = (v: number[]) => {
    const idx = v.map((val, i) => ({ val, i })).sort((a, b) => a.val - b.val);
    const r = new Array<number>(v.length);
    for (let i = 0; i < idx.length; ) {
      let j = i;
      while (j + 1 < idx.length && idx[j + 1].val === idx[i].val) j++;
      const avg = (i + j) / 2 + 1;
      for (let k = i; k <= j; k++) r[idx[k].i] = avg;
      i = j + 1;
    }
    return r;
  };
  const rx = rank(x), ry = rank(y);
  const mx = mean(rx), my = mean(ry);
  const num = rx.reduce((s, v, i) => s + (v - mx) * (ry[i] - my), 0);
  const den = Math.sqrt(rx.reduce((s, v) => s + (v - mx) ** 2, 0) * ry.reduce((s, v) => s + (v - my) ** 2, 0));
  return den === 0 ? 0 : num / den;
};
